mediapipe/graphs/instantmotiontracking/calculators/matrices_manager_calculator.cc

```diff
@@ -0,0 +1,324 @@
+// gl_animation_overlay_calculator. For information on the construction of OpenGL
+// objects and transformations (including a breakdown of model matrices), please
+// visit: https://open.gl/transformations
+//
+// Input Side Packets:
+//  FOV - Vertical field of view for device [REQUIRED - Defines perspective matrix]
+//  ASPECT_RATIO - Aspect ratio of device [REQUIRED - Defines perspective matrix]
+//
+// Input:
+//  ANCHORS - Anchor data with normalized x,y,z coordinates [REQUIRED]
+//  IMU_DATA - float[3] of [roll, pitch, yaw] of device [REQUIRED]
+//  USER_ROTATIONS - UserRotations with corresponding radians of rotation [REQUIRED]
+//  USER_SCALINGS - UserScalings with corresponding scale factor [REQUIRED]
+// Output:
+//  MODEL_MATRICES - TimedModelMatrixProtoList of all objects to render [REQUIRED]
+//
+// Example config:
+// node{
+//  calculator: "MatricesManagerCalculator"
+//  input_stream: "ANCHORS:tracked_scaled_anchor_data"
+//  input_stream: "IMU_DATA:imu_data"
+//  input_stream: "USER_ROTATIONS:user_rotation_data"
+//  input_stream: "USER_SCALINGS:user_scaling_data"
+//  output_stream: "MODEL_MATRICES:model_matrices"
+// }
+
+class MatricesManagerCalculator : public CalculatorBase {
+  public:
+    static ::mediapipe::Status GetContract(CalculatorContract* cc);
+    ::mediapipe::Status Open(CalculatorContext* cc) override;
+    ::mediapipe::Status Process(CalculatorContext* cc) override;
+  private:
+    const DiagonalMatrix3f GenerateScalingMatrix(const float scale_factor);
+    const Matrix3f GenerateUserRotationMatrix(const float rotation_radians);
+    const Matrix4fCM GenerateEigenModelMatrix(const Vector3f translation_vector,
+      const Matrix3f rotation_submatrix);
+    const Vector3f GenerateAnchorVector(const Anchor tracked_anchor);
+    const Matrix3f GenerateIMURotationSubmatrix(const float yaw, const float pitch, const float roll);
+
+    // Returns a user scaling increment associated with the sticker_id
+    // TODO: Adjust lookup function if total number of stickers is uncapped to improve performance
+    const float GetUserScaler(const std::vector<UserScaling> scalings, const int sticker_id) {
+      for (const UserScaling &user_scaling : scalings) {
+        if (user_scaling.sticker_id == sticker_id) {
+          return user_scaling.scale_factor;
+        }
+      }
+    }
+    // Returns a user rotation in radians associated with the sticker_id
+    const float GetUserRotation(const std::vector<UserRotation> rotations, const int sticker_id) {
+      for (const UserRotation &rotation : rotations) {
+        if (rotation.sticker_id == sticker_id) {
+          return rotation.rotation_radians;
+        }
+      }
+    }
+
+    // This returns a scale factor by which to alter the projection matrix for
+    // the specified render id in order to ensure all objects render at a similar
+    // size in the view screen upon initial placement
+    const float GetDefaultRenderScale(const int render_id) {
```

Comment:
Fine to leave as TODO for now, but eventually, initial sticker setup (default render scale, default rotation, default translation) should probably be baked into sticker asset or piped in / looked up as some sort of sticker config somewhere.